README.md
# TYPO3 Extension "jumpurl"

Adds functionality to configure and handle short links in the TYPO3 Frontend.

The functionality was part of TYPO3, until TYPO3 CMS 7 LTS, and moved into its own
extension, receiving its own public repository.

## Installation
The latest version can be installed via TER (http://typo3.org) or via composer
by adding ''composer require friendsoftypo3/jumpurl'' in a TYPO3 CMS 7+ installation.

## Current state
The latest version here reflects a feature-complete state.

## Contribution
Feel free to submit any pull request, or add documentation, tests, as you please.
We will publish a new version every once in a while, depending on the amount of changes
and pull requests submitted.

### License
The extension is published under GPL v2+, all included third-party libraries are
published under their respective licenses.

### Authors
Initially developed by Kasper Skaarhoj, it was polished by over 100 contributors in this
area while being part of the TYPO3 Core, and then extracted into its own extension by
Alexander Stehlik and Benni Mack.

This package is now maintained by a loose group of TYPO3 enthusiasts inside
the TYPO3 Community. Feel free to contact Benni Mack (benni.mack@typo3.org) for any questions
regarding "jumpurl".


ChangeLog
2017-11-30         [RELEASE] Release of jumpurl 7.7.1 (Nicole Cordes)
2017-11-23 9b1780b [BUGFIX] Fix unit tests (Nicole Cordes)
2017-11-23 94aa0b4 [FEATURE] Execute unit tests on Travis CI (Nicole Cordes)
2017-11-23 5f76ec7 [FEATURE] Add basic Travis CI configuration (Nicole Cordes)
2017-11-23 354f311 [TASK] Fix coding style according to php-cs-fixer (Nicole Cordes)
2017-11-23 ce7f713 [FEATURE] Add StyleCI configuration (Nicole Cordes)
2017-11-23 a47a604 [FEATURE] Include PHP CS Fixer configuration (Nicole Cordes)
2017-11-23 c87a8c6 [TASK] Add .gitattributes file (Nicole Cordes)
2017-11-23 5708d8d [TASK] Add .gitignore file (Nicole Cordes)
2017-10-18 9aa7cc6 Add correct replace properties (Alexander Grein)

2017-05-02         [RELEASE] Release of jumpurl 7.7.0
2017-05-02 224aa68 [TASK] Add support for TYPO3 v8 LTS (Benni Mack)
2015-11-03 ef1aa33 [TASK] Provide documentation Settings.cfg (Helmut Hummel)

2015-10-28         [RELEASE] Release of jumpurl 7.6.0
2015-10-28 c121859 [TASK] Added license information (Benjamin Mack)
2015-10-28 220c8db [TASK] Change PHP namespace of EXT:jumpurl to FoT3 (Benjamin Mack)
2015-10-09 a350127 [TASK] Make TYPO3 Core PSR-2 standard compliant (TYPO3 Release Team)
2015-09-29 7a2fd6f [TASK] Update TYPO3 version to TYPO3 7.6-dev (Wouter Wolters)
2015-09-23 d6a7f88 [BUGFIX] Adjust visibility of setUp() and tearDown() (Helmut Hummel)
2015-09-21 098d743 [TASK] Change typo3.org links to https (Josef Glatz)


Tests/Unit/JumpUrlHandlerTest.php
<?php
namespace FoT3\Jumpurl\Tests\Unit;

/*
 * This file is part of the TYPO3 CMS project.
 *
 * It is free software; you can redistribute it and/or modify it under
 * the terms of the GNU General Public License, either version 2
 * of the License, or any later version.
 *
 * For the full copyright and license information, please read the
 * LICENSE.txt file that was distributed with this source code.
 *
 * The TYPO3 project - inspiring people to share!
 */

use FoT3\Jumpurl\JumpUrlHandler;
use FoT3\Jumpurl\JumpUrlUtility;
use Nimut\TestingFramework\TestCase\UnitTestCase;
use TYPO3\CMS\Core\Resource\Exception\FileDoesNotExistException;
use TYPO3\CMS\Core\Resource\File;
use TYPO3\CMS\Core\Resource\ResourceFactory;
use TYPO3\CMS\Core\Utility\HttpUtility;
use TYPO3\CMS\Frontend\Authentication\FrontendUserAuthentication;
use TYPO3\CMS\Frontend\Controller\TypoScriptFrontendController;

/**
 * Testcase for handling jump URLs when given with a test parameter
 */
class JumpUrlHandlerTest extends UnitTestCase
{
    /**
     * The default location data used for JumpUrl secure.
     *
     * @var string
     */
    protected $defaultLocationData = '1234:tt_content:999';

    /**
     * @var \PHPUnit_Framework_MockObject_MockObject|JumpUrlHandler
     */
    protected $jumpUrlHandler;

    /**
     * @var \PHPUnit_Framework_MockObject_MockObject|\TYPO3\CMS\Core\Tests\AccessibleObjectInterface|\TYPO3\CMS\Frontend\Controller\TypoScriptFrontendController
     */
    protected $tsfe;

    /**
     * Sets environment variables and initializes global mock object.
     */
    protected function setUp()
    {
        parent::setUp();

        $GLOBALS['TYPO3_CONF_VARS']['SYS']['encryptionKey'] = '12345';

        $this->jumpUrlHandler = $this->getMock(
            JumpUrlHandler::class,
            ['isLocationDataValid', 'getResourceFactory', 'getTypoScriptFrontendController', 'readFileAndExit', 'redirect']
        );

        $this->tsfe = $this->getAccessibleMock(
            TypoScriptFrontendController::class,
            ['getPagesTSconfig'],
            [],
            '',
            false
        );
        $this->jumpUrlHandler->expects($this->any())
            ->method('getTypoScriptFrontendController')
            ->will($this->returnValue($this->tsfe));
    }

    /**
     * Provides a valid jump URL hash and a target URL
     *
     * @return array
     */
    public function jumpUrlDefaultValidParametersDataProvider()
    {
        return [
            'File with spaces and ampersands' => [
                '691dbf63a21181e2d69bf78e61f1c9fd023aef2c',
                str_replace('%2F', '/', rawurlencode('typo3temp/phpunitJumpUrlTestFile with spaces & amps.txt')),
            ],
            'External URL' => [
                '7d2261b12682a4b73402ae67415e09f294b29a55',
                'http://www.mytesturl.tld',
            ],
            'External URL with GET parameters' => [
                'cfc95f583da7689238e98bbc8930ebd820f0d20f',
                'http://external.domain.tld?parameter1=' . rawurlencode('parameter[data]with&a lot-of-special/chars'),
            ],
            'External URL without www' => [
                '8591c573601d17f37e06aff4ac14c78f107dd49e',
                'http://external.domain.tld',
            ],
            'Mailto link' => [
                'bd82328dc40755f5d0411e2e16e7c0cbf33b51b7',
                'mailto:mail@ddress.tld',
            ]
        ];
    }

    /**
     * @test
     * @dataProvider jumpUrlDefaultValidParametersDataProvider
     * @param string $hash
     * @param string $jumpUrl
     */
    public function jumpUrlDefaultAcceptsValidUrls($hash, $jumpUrl)
    {
        $_GET['juHash'] = $hash;
        $_GET['jumpurl'] = $jumpUrl;

        $this->jumpUrlHandler->expects($this->once())
            ->method('redirect')
            ->with($jumpUrl, HttpUtility::HTTP_STATUS_303);

        $this->jumpUrlHandler->canHandleCurrentUrl();
        $this->jumpUrlHandler->handle();
    }

    /**
     * @test
     * @dataProvider jumpUrlDefaultValidParametersDataProvider
     * @expectedException \Exception
     * @expectedExceptionCode 1359987599
     * @param string $hash
     * @param string $jumpUrl
     */
    public function jumpUrlDefaultFailsOnInvalidHash($hash, $jumpUrl)
    {
        $_GET['jumpurl'] = $jumpUrl;
        $_GET['juHash'] = $hash . '1';

        $this->jumpUrlHandler->canHandleCurrentUrl();
        $this->jumpUrlHandler->handle();
    }

    /**
     * @test
     * @dataProvider jumpUrlDefaultValidParametersDataProvider
     * @param string $hash
     * @param string $jumpUrl
     */
    public function jumpUrlDefaultTransfersSession($hash, $jumpUrl)
    {
        $tsConfig['TSFE.']['jumpUrl_transferSession'] = 1;

        /** @var \PHPUnit_Framework_MockObject_MockObject|FrontendUserAuthentication $frontendUserMock */
        $frontendUserMock = $this->getMock(FrontendUserAuthentication::class);
        $frontendUserMock->id = 123;

        $this->tsfe->_set('fe_user', $frontendUserMock);
        $this->tsfe->expects($this->once())
            ->method('getPagesTSconfig')
            ->will($this->returnValue($tsConfig));

        $sessionGetParameter = (strpos($jumpUrl, '?') === false ? '?' : '') . '&FE_SESSION_KEY=123-fc9f825a9af59169895f3bb28267a42f';
        $expectedJumpUrl = $jumpUrl . $sessionGetParameter;

        $this->jumpUrlHandler->expects($this->once())
            ->method('redirect')
            ->with($expectedJumpUrl, HttpUtility::HTTP_STATUS_303);

        $_GET['jumpurl'] = $jumpUrl;
        $_GET['juHash'] = $hash;

        $this->jumpUrlHandler->canHandleCurrentUrl();
        $this->jumpUrlHandler->handle();
    }

    /**
     * Provides a valid jump secure URL hash, a file path and related
     * record data
     *
     * @return array
     */
    public function jumpUrlSecureValidParametersDataProvider()
    {
        return [
            [
                '1933f3c181db8940acfcd4d16c74643947179948',
                'typo3temp/phpunitJumpUrlTestFile.txt',
            ],
            [
                '304b8c8e022e92e6f4d34e97395da77705830818',
                str_replace('%2F', '/', rawurlencode('typo3temp/phpunitJumpUrlTestFile with spaces & amps.txt')),
            ],
            [
                '304b8c8e022e92e6f4d34e97395da77705830818',
                str_replace('%2F', '/', rawurlencode('typo3temp/phpunitJumpUrlTestFile with spaces & amps.txt')),
            ]
        ];
    }

    /**
     * @test
     * @dataProvider jumpUrlSecureValidParametersDataProvider
     * @param string $hash
     * @param string $jumpUrl
     */
    public function jumpUrlSecureAcceptsValidUrls($hash, $jumpUrl)
    {
        $_GET['jumpurl'] = $jumpUrl;
        $this->prepareJumpUrlSecureTest($hash);

        $fileMock = $this->getMock(File::class, ['dummy'], [], '', false);
        $resourceFactoryMock = $this->getMock(ResourceFactory::class, ['retrieveFileOrFolderObject']);

        $resourceFactoryMock->expects($this->once())
            ->method('retrieveFileOrFolderObject')
            ->will($this->returnValue($fileMock));

        $this->jumpUrlHandler->expects($this->once())
            ->method('isLocationDataValid')
            ->with($this->defaultLocationData)
            ->will($this->returnValue(true));

        $this->jumpUrlHandler->expects($this->once())
            ->method('getResourceFactory')
            ->will($this->returnValue($resourceFactoryMock));

        $this->jumpUrlHandler->expects($this->once())
            ->method('readFileAndExit')
            ->with($fileMock);

        $this->jumpUrlHandler->canHandleCurrentUrl();
        $this->jumpUrlHandler->handle();
    }

    /**
     * @test
     * @dataProvider jumpUrlSecureValidParametersDataProvider
     * @expectedException \Exception
     * @expectedExceptionCode 1294585193
     * @param string $hash
     * @param string $jumpUrl
     */
    public function jumpUrlSecureFailsIfFileDoesNotExist($hash, $jumpUrl)
    {
        $_GET['jumpurl'] = $jumpUrl;
        $this->prepareJumpUrlSecureTest($hash);

        $resourceFactoryMock = $this->getMock(ResourceFactory::class, ['retrieveFileOrFolderObject']);
        $resourceFactoryMock->expects($this->once())
            ->method('retrieveFileOrFolderObject')
            ->will($this->throwException(new FileDoesNotExistException()));

        $this->jumpUrlHandler->expects($this->once())
            ->method('isLocationDataValid')
            ->with($this->defaultLocationData)
            ->will($this->returnValue(true));

        $this->jumpUrlHandler->expects($this->once())
            ->method('getResourceFactory')
            ->will($this->returnValue($resourceFactoryMock));

        $this->jumpUrlHandler->canHandleCurrentUrl();
        $this->jumpUrlHandler->handle();
    }

    /**
     * @test
     * @dataProvider jumpUrlSecureValidParametersDataProvider
     * @expectedException \Exception
     * @expectedExceptionCode 1294585195
     * @param string $hash
     * @param string $jumpUrl
     */
    public function jumpUrlSecureFailsOnDeniedAccess($hash, $jumpUrl)
    {
        $_GET['jumpurl'] = $jumpUrl;
        $this->prepareJumpUrlSecureTest($hash);

        $this->jumpUrlHandler->expects($this->once())
            ->method('isLocationDataValid')
            ->with($this->defaultLocationData)
            ->will($this->returnValue(false));

        $this->jumpUrlHandler->canHandleCurrentUrl();
        $this->jumpUrlHandler->handle();
    }

    /**
     * @test
     * @dataProvider jumpUrlSecureValidParametersDataProvider
     * @expectedException \Exception
     * @expectedExceptionCode 1294585196
     * @param string $hash
     * @param string $jumpUrl
     */
    public function jumpUrlSecureFailsOnInvalidHash(
        $hash,
        $jumpUrl
    ) {
        $_GET['juSecure'] = '1';
        $_GET['juHash'] = $hash . '1';
        $_GET['locationData'] = $this->defaultLocationData;

        $this->jumpUrlHandler->canHandleCurrentUrl();
        $this->jumpUrlHandler->handle();
    }

    /**
     * @return array
     */
    public function jumpUrlSecureFailsOnForbiddenFileLocationDataProvider()
    {
        return [
            'totally forbidden' => [
                '/a/totally/forbidden/path'
            ],
            'typo3conf file' => [
                PATH_site . '/typo3conf/path'
            ],
            'file with forbidden character' => [
                PATH_site . '/mypath/test.php'
            ]
        ];
    }

    /**
     * @test
     * @dataProvider jumpUrlSecureFailsOnForbiddenFileLocationDataProvider
     * @expectedException \Exception
     * @expectedExceptionCode 1294585194
     * @param string $path
     */
    public function jumpUrlSecureFailsOnForbiddenFileLocation($path)
    {
        $this->jumpUrlHandler->expects($this->once())
            ->method('isLocationDataValid')
            ->with('')
            ->will($this->returnValue(true));

        $hash = JumpUrlUtility::calculateHashSecure($path, '', '');

        $_GET['jumpurl'] = $path;
        $_GET['juSecure'] = '1';
        $_GET['juHash'] = $hash;
        $_GET['locationData'] = '';

        $this->jumpUrlHandler->canHandleCurrentUrl();
        $this->jumpUrlHandler->handle();
    }

    /**
     * @param string $hash
     */
    protected function prepareJumpUrlSecureTest($hash)
    {
        $_GET['juSecure'] = '1';
        $_GET['juHash'] = $hash;
        $_GET['locationData'] = $this->defaultLocationData;
    }
}


Tests/Unit/JumpUrlProcessorTest.php
<?php
namespace FoT3\Jumpurl\Tests\Unit;

/*
 * This file is part of the TYPO3 CMS project.
 *
 * It is free software; you can redistribute it and/or modify it under
 * the terms of the GNU General Public License, either version 2
 * of the License, or any later version.
 *
 * For the full copyright and license information, please read the
 * LICENSE.txt file that was distributed with this source code.
 *
 * The TYPO3 project - inspiring people to share!
 */

use Nimut\TestingFramework\TestCase\UnitTestCase;
use TYPO3\CMS\Frontend\ContentObject\ContentObjectRenderer;
use TYPO3\CMS\Frontend\Controller\TypoScriptFrontendController;

/**
 * Testcase for handling jump URLs when given with a test parameter
 */
class JumpUrlProcessorTest extends UnitTestCase
{
    /**
     * @var \PHPUnit_Framework_MockObject_MockObject|ContentObjectRenderer
     */
    protected $contentObjectRenderer;

    /**
     * The default location data used for JumpUrl secure.
     *
     * @var string
     */
    protected $defaultLocationData = '1234:tt_content:999';

    /**
     * @var \PHPUnit_Framework_MockObject_MockObject|JumpUrlProcessorMock
     */
    protected $jumpUrlProcessor;

    /**
     * @var \PHPUnit_Framework_MockObject_MockObject|\TYPO3\CMS\Core\Tests\AccessibleObjectInterface|\TYPO3\CMS\Frontend\Controller\TypoScriptFrontendController
     */
    protected $tsfe;

    /**
     * Sets environment variables and initializes global mock object.
     */
    protected function setUp()
    {
        parent::setUp();

        $GLOBALS['TYPO3_CONF_VARS']['SYS']['encryptionKey'] = '12345';

        $this->jumpUrlProcessor = $this->getMock(
            JumpUrlProcessorMock::class,
            ['getTypoScriptFrontendController', 'getContentObjectRenderer']
        );

        $this->tsfe = $this->getAccessibleMock(
            TypoScriptFrontendController::class,
            ['getPagesTSconfig'],
            [],
            '',
            false
        );
        $this->jumpUrlProcessor->expects($this->any())
            ->method('getTypoScriptFrontendController')
            ->will($this->returnValue($this->tsfe));

        $this->contentObjectRenderer = $this->getMock(ContentObjectRenderer::class);
        $this->jumpUrlProcessor->expects($this->any())
            ->method('getContentObjectRenderer')
            ->will($this->returnValue($this->contentObjectRenderer));
    }

    /**
     * @test
     */
    public function getJumpUrlSecureParametersReturnsValidParameters()
    {
        $this->tsfe->id = 456;
        $this->contentObjectRenderer->currentRecord = 'tt_content:123';

        $jumpUrlSecureParameters = $this->jumpUrlProcessor->getParametersForSecureFile(
            '/fileadmin/a/test/file.txt',
            ['mimeTypes' => 'dummy=application/x-executable,txt=text/plain']
        );

        $this->assertSame(
            [
                'juSecure' => 1,
                'locationData' => '456:tt_content:123',
                'mimeType' => 'text/plain',
                'juHash' => '1cccb7f01c8a3f58ee890377b5de9bdc05115a37',
            ],
            $jumpUrlSecureParameters
        );
    }
}


Classes/JumpUrlUtility.php
<?php
namespace FoT3\Jumpurl;

/*
 * This file is part of the TYPO3 CMS project.
 *
 * It is free software; you can redistribute it and/or modify it under
 * the terms of the GNU General Public License, either version 2
 * of the License, or any later version.
 *
 * For the full copyright and license information, please read the
 * LICENSE.txt file that was distributed with this source code.
 *
 * The TYPO3 project - inspiring people to share!
 */

use TYPO3\CMS\Core\Utility\GeneralUtility;

/**
 * This class contains functions for generating and validating jump URLs
 */
class JumpUrlUtility
{
    /**
     * Calculates the hash for the given jump URL
     *
     * @param string $jumpUrl The target URL
     * @return string The calculated hash
     */
    public static function calculateHash($jumpUrl)
    {
        return GeneralUtility::hmac($jumpUrl, 'jumpurl');
    }

    /**
     * Calculates the hash for the given jump URL secure data.
     *
     * @param string $jumpUrl The URL to the file
     * @param string $locationData Information about the record that rendered the jump URL, format is [pid]:[table]:[uid]
     * @param string $mimeType Mime type of the file or an empty string
     * @return string The calculated hash
     */
    public static function calculateHashSecure($jumpUrl, $locationData, $mimeType)
    {
        $data = [(string)$jumpUrl, (string)$locationData, (string)$mimeType];
        return GeneralUtility::hmac(serialize($data));
    }
}


Classes/JumpUrlHandler.php
<?php
namespace FoT3\Jumpurl;

/*
 * This file is part of the TYPO3 CMS project.
 *
 * It is free software; you can redistribute it and/or modify it under
 * the terms of the GNU General Public License, either version 2
 * of the License, or any later version.
 *
 * For the full copyright and license information, please read the
 * LICENSE.txt file that was distributed with this source code.
 *
 * The TYPO3 project - inspiring people to share!
 */

use TYPO3\CMS\Core\Resource\FileInterface;
use TYPO3\CMS\Core\Resource\ResourceFactory;
use TYPO3\CMS\Core\Utility\GeneralUtility;
use TYPO3\CMS\Core\Utility\HttpUtility;
use TYPO3\CMS\Frontend\Http\UrlHandlerInterface;

/**
 * This class implements the hooks for the JumpURL functionality when accessing a page
 * which has a GET parameter "jumpurl".
 * It then validates the referrer
 */
class JumpUrlHandler implements UrlHandlerInterface
{
    /**
     * @var string The current JumpURL value submitted in the GET parameters.
     */
    protected $url;

    /**
     * Return TRUE if this hook handles the current URL.
     * Warning! If TRUE is returned content rendering will be disabled!
     * This method will be called in the constructor of the TypoScriptFrontendController
     *
     * @see \TYPO3\CMS\Frontend\Controller\TypoScriptFrontendController::__construct()
     * @see \TYPO3\CMS\Frontend\Controller\TypoScriptFrontendController::initializeCustomUrlHandlers()
     * @return bool
     */
    public function canHandleCurrentUrl()
    {
        $this->url = (string)GeneralUtility::_GP('jumpurl');
        return $this->url !== '';
    }

    /**
     * Custom processing of the current URL.
     *
     * If a valid hash was submitted the user will either be redirected
     * to the given jumpUrl or if it is a secure jumpUrl the file data
     * will be passed to the user.
     *
     * If canHandle() has returned TRUE this method needs to take care of redirecting the user or generating custom output.
     * This hook will be called BEFORE the user is redirected to an external URL configured in the page properties.
     *
     * @see \TYPO3\CMS\Frontend\Controller\TypoScriptFrontendController::processCustomUrlHandlers()
     * @throws \RuntimeException if Jump URL was triggered by an illegal referrer.
     */
    public function handle()
    {
        if (!$this->referrerIsValid()) {
            throw new \RuntimeException('The jumpUrl request was triggered by an illegal referrer.');
        }

        if ((bool)GeneralUtility::_GP('juSecure')) {
            $this->forwardJumpUrlSecureFileData($this->url);
        } else {
            $this->redirectToJumpUrl($this->url);
        }
    }

    /**
     * Returns TRUE if the current referrer allows Jump URL handling.
     * This is the case then the referrer check is disabled or when the referrer matches the current TYPO3 host.
     *
     * @return bool if the referer is valid.
     */
    protected function referrerIsValid()
    {
        if (!empty($GLOBALS['TYPO3_CONF_VARS']['SYS']['doNotCheckReferer'])) {
            return true;
        }

        $referrer = parse_url(GeneralUtility::getIndpEnv('HTTP_REFERER'));
        // everything is fine if no host is set, or the host matches the TYPO3_HOST
        return !isset($referrer['host']) || $referrer['host'] === GeneralUtility::getIndpEnv('TYPO3_HOST_ONLY');
    }

    /**
     * Redirects the user to the given jump URL if all submitted values
     * are valid
     *
     * @param string $jumpUrl The URL to which the user should be redirected
     * @throws \Exception
     */
    protected function redirectToJumpUrl($jumpUrl)
    {
        $this->validateIfJumpUrlRedirectIsAllowed($jumpUrl);

        $pageTSconfig = $this->getTypoScriptFrontendController()->getPagesTSconfig();
        if (is_array($pageTSconfig['TSFE.'])) {
            $pageTSconfig = $pageTSconfig['TSFE.'];
        } else {
            $pageTSconfig = [];
        }

        $jumpUrl = $this->addParametersToTransferSession($jumpUrl, $pageTSconfig);
        $statusCode = $this->getRedirectStatusCode($pageTSconfig);
        $this->redirect($jumpUrl, $statusCode);
    }

    /**
     * If the submitted hash is correct and the user has access to the
     * related content element the contents of the submitted file will
     * be output to the user.
     *
     * @param string $jumpUrl The URL to the file that should be output to the user
     * @throws \Exception
     */
    protected function forwardJumpUrlSecureFileData($jumpUrl)
    {
        // Set the parameters required for handling a secure jumpUrl link
        // The locationData GET parameter, containing information about the record that created the URL
        $locationData = (string)GeneralUtility::_GP('locationData');
        // The optional mimeType GET parameter
        $mimeType = (string)GeneralUtility::_GP('mimeType');
        // The jump Url Hash GET parameter
        $juHash = (string)GeneralUtility::_GP('juHash');

        // validate the hash GET parameter against the other parameters
        if ($juHash !== JumpUrlUtility::calculateHashSecure($jumpUrl, $locationData, $mimeType)) {
            throw new \Exception('The calculated Jump URL secure hash ("juHash") did not match the submitted "juHash" query parameter.', 1294585196);
        }

        if (!$this->isLocationDataValid($locationData)) {
            throw new \Exception('The calculated secure location data "' . $locationData . '" is not accessible.', 1294585195);
        }

        // Allow spaces / special chars in filenames.
        $jumpUrl = rawurldecode($jumpUrl);

        // Deny access to files that match TYPO3_CONF_VARS[SYS][fileDenyPattern] and whose parent directory
        // is typo3conf/ (there could be a backup file in typo3conf/ which does not match against the fileDenyPattern)
        $absoluteFileName = GeneralUtility::getFileAbsFileName(GeneralUtility::resolveBackPath($jumpUrl), false);

        if (
            !GeneralUtility::isAllowedAbsPath($absoluteFileName)
            || !GeneralUtility::verifyFilenameAgainstDenyPattern($absoluteFileName)
            || GeneralUtility::isFirstPartOfStr($absoluteFileName, (PATH_site . 'typo3conf'))
        ) {
            throw new \Exception('The requested file was not allowed to be accessed through Jump URL. The path or file is not allowed.', 1294585194);
        }

        try {
            $resourceFactory = $this->getResourceFactory();
            $file = $resourceFactory->retrieveFileOrFolderObject($absoluteFileName);
            $this->readFileAndExit($file, $mimeType);
        } catch (\Exception $e) {
            throw new \Exception('The requested file "' . $jumpUrl . '" for Jump URL was not found..', 1294585193);
        }
    }

    /**
     * Checks if the given location data is valid and the connected record is accessible by the current user.
     *
     * @param string $locationData
     * @return bool
     */
    protected function isLocationDataValid($locationData)
    {
        $isValidLocationData = false;
        list($pageUid, $table, $recordUid) = explode(':', $locationData);
        $pageRepository = $this->getTypoScriptFrontendController()->sys_page;
        $timeTracker = $this->getTimeTracker();
        if (empty($table) || $pageRepository->checkRecord($table, $recordUid, true)) {
            // This check means that a record is checked only if the locationData has a value for a
            // record else than the page.
            if (!empty($pageRepository->getPage($pageUid))) {
                $isValidLocationData = true;
            } else {
                $timeTracker->setTSlogMessage('LocationData Error: The page pointed to by location data "' . $locationData . '" was not accessible.', 2);
            }
        } else {
            $timeTracker->setTSlogMessage('LocationData Error: Location data "' . $locationData . '" record pointed to was not accessible.', 2);
        }
        return $isValidLocationData;
    }

    /**
     * This implements a hook, e.g. for direct mail to allow the redirects but only if the handler says it's alright
     * But also checks against the common juHash parameter first
     *
     * @param string $jumpUrl the URL to check
     * @throws \Exception thrown if no redirect is allowed
     */
    protected function validateIfJumpUrlRedirectIsAllowed($jumpUrl)
    {
        $allowRedirect = false;
        if ($this->isJumpUrlHashValid($jumpUrl)) {
            $allowRedirect = true;
        } elseif (
            isset($GLOBALS['TYPO3_CONF_VARS']['SC_OPTIONS']['tslib/class.tslib_fe.php']['jumpurlRedirectHandler'])
            && is_array($GLOBALS['TYPO3_CONF_VARS']['SC_OPTIONS']['tslib/class.tslib_fe.php']['jumpurlRedirectHandler'])
        ) {
            foreach ($GLOBALS['TYPO3_CONF_VARS']['SC_OPTIONS']['tslib/class.tslib_fe.php']['jumpurlRedirectHandler'] as $className) {
                $hookObject = GeneralUtility::getUserObj($className);
                if (method_exists($hookObject, 'jumpurlRedirectHandler')) {
                    $allowRedirect = $hookObject->jumpurlRedirectHandler($jumpUrl, $GLOBALS['TSFE']);
                }
                if ($allowRedirect) {
                    break;
                }
            }
        }

        if (!$allowRedirect) {
            throw new \Exception('The calculated Jump URL hash ("juHash") did not match the submitted "juHash" query parameter.', 1359987599);
        }
    }

    /**
     * Validate the jumpUrl hash against the GET/POST parameter "juHash".
     *
     * @param string $jumpUrl The URL to check against.
     * @return bool
     */
    protected function isJumpUrlHashValid($jumpUrl)
    {
        return GeneralUtility::_GP('juHash') === JumpUrlUtility::calculateHash($jumpUrl);
    }

    /**
     * Calls the PHP readfile function and exits.
     *
     * @param FileInterface $file The file that should be read.
     * @param string $mimeType Optional mime type override. If empty the automatically detected mime type will be used.
     */
    protected function readFileAndExit($file, $mimeType)
    {
        $file->getStorage()->dumpFileContents($file, true, null, $mimeType);
        exit;
    }

    /**
     * Simply calls the redirect method in the HttpUtility.
     *
     * @param string $jumpUrl
     * @param int $statusCode
     */
    protected function redirect($jumpUrl, $statusCode)
    {
        HttpUtility::redirect($jumpUrl, $statusCode);
    }

    /**
     * Modified the URL to go to by adding the session key information to it
     * but only if TSFE.jumpUrl_transferSession = 1 is set via pageTSconfig.
     *
     * @param string $jumpUrl the URL to go to
     * @param array $pageTSconfig the TSFE. part of the TS configuration
     *
     * @return string the modified URL
     */
    protected function addParametersToTransferSession($jumpUrl, $pageTSconfig)
    {
        // allow to send the current fe_user with the jump URL
        if (!empty($pageTSconfig['jumpUrl_transferSession'])) {
            $uParts = parse_url($jumpUrl);
            /** @noinspection PhpInternalEntityUsedInspection We need access to the current frontend user ID. */
            $params = '&FE_SESSION_KEY=' .
                rawurlencode(
                    $this->getTypoScriptFrontendController()->fe_user->id . '-' .
                    md5(
                        $this->getTypoScriptFrontendController()->fe_user->id . '/' .
                        $GLOBALS['TYPO3_CONF_VARS']['SYS']['encryptionKey']
                    )
                );
            // Add the session parameter ...
            $jumpUrl .= ($uParts['query'] ? '' : '?') . $params;
        }
        return $jumpUrl;
    }

    /**
     * Returns one of the HTTP_STATUS_* constants of the HttpUtility that matches
     * the configured HTTP status code in TSFE.jumpURL_HTTPStatusCode Page TSconfig.
     *
     * @param array $pageTSconfig
     * @return string
     * @throws \InvalidArgumentException If the configured status code is not valid.
     */
    protected function getRedirectStatusCode($pageTSconfig)
    {
        $statusCode = HttpUtility::HTTP_STATUS_303;

        if (!empty($pageTSconfig['jumpURL_HTTPStatusCode'])) {
            switch ((int)$pageTSconfig['jumpURL_HTTPStatusCode']) {
                case 301:
                    $statusCode = HttpUtility::HTTP_STATUS_301;
                    break;
                case 302:
                    $statusCode = HttpUtility::HTTP_STATUS_302;
                    break;
                case 307:
                    $statusCode = HttpUtility::HTTP_STATUS_307;
                    break;
                default:
                    throw new \InvalidArgumentException('The configured jumpURL_HTTPStatusCode option is invalid. Allowed codes are 301, 302 and 307.', 1381768833);
            }
        }

        return $statusCode;
    }

    /**
     * @return \TYPO3\CMS\Core\TimeTracker\TimeTracker
     */
    protected function getTimeTracker()
    {
        return $GLOBALS['TT'];
    }

    /**
     * Returns an instance of the ResourceFactory.
     *
     * @return ResourceFactory
     */
    protected function getResourceFactory()
    {
        return ResourceFactory::getInstance();
    }

    /**
     * @return \TYPO3\CMS\Frontend\Controller\TypoScriptFrontendController
     */
    protected function getTypoScriptFrontendController()
    {
        return $GLOBALS['TSFE'];
    }
}
